lib/spreewald_support/mail_finder.rb

```diff
@@ -5,22 +5,38 @@ class << self
 
     attr_accessor :user_identity
 
-    def find(conditions)
+    def find(raw_data, type = '')
+      header, body = raw_data.split(/\n\n/, 2) # 2: maximum number of fields
+      conditions = {}
+      header.split("\n").each do |row|
+        if row.match(/^[A-Za-z\-]+: /i)
+          key, value = row.split(": ", 2)
+          conditions[key.strip.underscore.to_sym] = value.strip
+        end
+      end
+
       filename_method = Rails::VERSION::MAJOR < 3 ? 'original_filename' : 'filename'
-      ActionMailer::Base.deliveries.detect do |mail|
-        mail_body = email_text_body(mail)
+      matching_header = ActionMailer::Base.deliveries.select do |mail|
         [ conditions[:to].nil? || mail.to.include?(resolve_email conditions[:to]),
           conditions[:cc].nil? || mail.cc && mail.cc.include?(resolve_email conditions[:cc]),
           conditions[:bcc].nil? || mail.bcc && mail.bcc.include?(resolve_email conditions[:bcc]),
           conditions[:from].nil? || mail.from.include?(resolve_email conditions[:from]),
           conditions[:reply_to].nil? || mail.reply_to.include?(resolve_email conditions[:reply_to]),
           conditions[:subject].nil? || mail.subject.include?(conditions[:subject]),
-          conditions[:body].nil? || mail_body.include?(conditions[:body]),
           conditions[:attachments].nil? || conditions[:attachments].split(/\s*,\s*/).sort == Array(mail.attachments).collect(&:"#{filename_method}").sort
         ].all?
-      end.tap do |mail|
-        log(mail)
       end
+
+      matching_mails = if body
+        body_regex = expected_body_regex(body)
+        matching_header.select do |mail|
+          body_regex =~ email_text_body(mail, type).strip
+        end
+      else
+        matching_header
+      end
+
+      {matching: matching_mails, matching_header: matching_header, body_regex: body_regex}
```

Comment:
All steps using the MailFinder have to deal with this Hash structure. This feels weird in an object-oriented world and is error-prone when calling non-existent keys. Instead, provide an object.

That way, the object may also implement logic and steps do not have to know about structural internals.

Example (I like using `Struct`, not a requirement):
```ruby
class MailFinder::Results < Struct.new(:matching, :matching_header, :body_regex)
  extend Forwardable
  delegate [:size, :one?, :empty?] => :matching

  def many?
    size == 2
  end
end
```

`MailFinder` would then instead return a `Results` instance:
```suggestion
      Results.new(matching, matching_header, body_regex)
```

Example usage:
`expect(mail[:matching]).to be_empty` => `expect(mail).to be_empty`
`mail[:matching].size == 1` => `mail.one?`
`mail[:matching].size > 1` => `mail.many?`

I'd also suggest calling the variable in steps `mails` or `results`. After all, there may be multiple.
When you call it `results`, consider renaming `matching` to something more human-understandable like `mails` (would lead to `results.mails`).